### PR TITLE
Move `is_data_available` check to fork-choice `on_block`

### PR DESCRIPTION
**Pull request by hwwhww**
Address #3170

Note: need to check again the confidence level of this change.

### CODE REVIEW

File: specs/eip4844/fork-choice.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+This is the modification of the fork choice according to the executable beacon chain proposal.
+
+## Helpers
+
+#### `is_data_available`
+
+The implementation of `is_data_available` will become more sophisticated during later sharding upgrades.
+Initially, it requires every verifying actor to retrieve the matching `BlobsSidecar`,
+and validate the sidecar with `validate_blobs_sidecar`.
```
Comment (djrtwo) [inline specs/eip4844/fork-choice.md:27]: Do you think we should move `validate_blobs_sidecar` here as well? p2p spec also uses that function but I *think* it's more appropriate here closer to `is_data_available`

On second thought, the dependency graph between the files is a bit strange right now:
* `class BlobsSidecar` defined in p2p spec and p2p spec references `validate_blobs_sidecar()`
* `validate_blobs_sidecar()` defined in beacon chain spec and references `BlobsSideCar`
* `is_data_available()` defined in fork-choice spec and references side ccars and `validate_blobs_sidecar`

I *think* I'd move all of it to the fork-choice and have the fork-choice doc be a clean dependency for both beacon chain and p2p spec

File: specs/eip4844/fork-choice.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    assert get_current_slot(store) >= block.slot
+
+    # Check that block is later than the finalized epoch slot (optimization to reduce calls to get_ancestor)
+    finalized_slot = compute_start_slot_at_epoch(store.finalized_checkpoint.epoch)
+    assert block.slot > finalized_slot
+    # Check block is a descendant of the finalized block at the checkpoint finalized slot
+    assert get_ancestor(store, block.parent_root, finalized_slot) == store.finalized_checkpoint.root
+
+    # [New in EIP-4844]
+    # Check if the block is available
```
Comment (djrtwo) [inline specs/eip4844/fork-choice.md:72]: I think we should make explicit that the block can be considered in the future if the blobs become available.
(kind of like if you got a block too early, you delay consideration until it's in the past)

```suggestion
    # Check if blob data is available
    # If not, this block MAY be queued and subsequently considered when blob data becomes available
```

**Review by djrtwo — COMMENTED** [review]
I left some relatively minor comments on file/definition organization

Also, the fact that we have no tests change here is a bit unnerving -- seeming we don't have any tests for `is_data_available()` within fork-choice or state transition.

It might be good to upgtrade the fork-choice test format to include if a block is available. (and if not, be able to subsequently become available) but I'm not sure if we should do so here

File: specs/eip4844/fork-choice.md (hunk 1 above)
Comment (hwwhww) [reply specs/eip4844/fork-choice.md:27]: > move all of it to the fork-choice and have the fork-choice doc be a clean dependency for both beacon chain and p2p spec

Agreed. It makes more sense according to the dependencies.

Comment (djrtwo) [reply specs/eip4844/fork-choice.md:44]: What about the `class BlobSidecar` def? I think it should go in here as well

**Review by djrtwo — APPROVED** [review]
excellent!

**Comment by hwwhww** [comment]
> It might be good to upgtrade the fork-choice test format to include if a block is available. (and if not, be able to subsequently become available) but I'm not sure if we should do so here

We could do something like the way we hacked `get_pow_block` in `on_merge_block` tests: https://github.com/ethereum/consensus-specs/blob/e9a7a6d40ee853dc2ec3dbcf3b2e9d4f5c6b6c27/tests/core/pyspec/eth2spec/test/bellatrix/fork_choice/test_on_merge_block.py#L29-L50

I'll open a ticket, but probably add it in the next next release?

**Pull request merged**